sources/api/bootstrap-containers/src/main.rs

```diff
@@ -0,0 +1,661 @@
+
+```toml
+[settings.bootstrap-containers.bear]
+source="<SOURCE>"
+mode="once"
+user-data="ypXCt82h4bSlwrfKlA=="
+```
+
+Where `<SOURCE>`, is the url of an image with the following definition:
+
+```Dockerfile
+FROM alpine
+ADD bootstrap-script /
+RUN chmod +x /bootstrap-script
+ENTRYPOINT ["sh", "bootstrap-script"]
+```
+
+And `bootstrap-script` as:
+
+```sh
+#!/bin/sh
+BEAR_DIR=/.bottlerocket/bootstrap-containers/bear
+VAR_DIR=/.bottlerocket/rootfs/var
+MY_DIR=$VAR_DIR/lib/my_directory
+cat $BEAR_DIR/user-data
+mkdir $MY_DIR
+chmod -R o+r $MY_DIR
+chown -R 1000:1000 $MY_DIR
+sleep 120
+```
+
+You should see a new directory under `/var/lib` called `my_directory`, and
+the following command should show something similar to `ʕ·͡ᴥ·ʔ`:
+
+```sh
+journalctl -u bootstrap-containers@bear.service
+```
+*/
+
+#![deny(rust_2018_idioms)]
+
+#[macro_use]
+extern crate log;
+
+use datastore::{serialize_scalar, Key, KeyType};
+use simplelog::{Config as LogConfig, LevelFilter, SimpleLogger};
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::env;
+use std::ffi::OsStr;
+use std::fmt::Write;
+use std::fs;
+use std::path::Path;
+use std::process::{self, Command};
+use std::str::FromStr;
+
+use model::modeled_types::{Identifier, ValidContainerMode};
+
+// FIXME Get from configuration in the future
+const DEFAULT_API_SOCKET: &str = "/run/api.sock";
+const API_SETTINGS_URI: &str = "/settings";
+
+const ENV_FILE_DIR: &str = "/etc/bootstrap-containers";
+const PERSISTENT_STORAGE_DIR: &str = "/local/bootstrap-containers";
+
+const SYSTEMCTL_BIN: &str = "/bin/systemctl";
+const HOST_CTR_BIN: &str = "/bin/host-ctr";
+
+/// Stores user-supplied global arguments
+#[derive(Debug)]
+struct Args {
+    log_level: LevelFilter,
+    socket_path: String,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+            socket_path: DEFAULT_API_SOCKET.to_string(),
+        }
+    }
+}
+
+/// Stores the subcommand to be executed
+#[derive(Debug)]
+enum Subcommand {
+    CreateContainers,
+    MarkBootstrap(MarkBootstrapArgs),
+}
+
+#[derive(Debug)]
+struct MarkBootstrapArgs {
+    container_id: String,
+    mode: ValidContainerMode,
+}
+
+/// Print a usage message in the event a bad arg is passed
+fn usage() {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!(
+        r"Usage: {} SUBCOMMAND [ ARGUMENTS... ]
+
+    Subcommands:
+        create-containers
+        mark-bootstrap
+
+    Global arguments:
+        [ --socket-path PATH ]
+        [ --log-level trace|debug|info|warn|error ]
+
+    Mark bootstrap arguments:
+        --container-id CONTAINER-ID
+        --mode MODE
+
+    Socket path defaults to {}",
+        program_name, DEFAULT_API_SOCKET,
+    );
+}
+
+/// Parses user arguments into an Args struct
+fn parse_args(args: env::Args) -> Result<(Args, Subcommand)> {
```

Comment:
I would expect the Subcommand to be a member of the args struct.

```suggestion
fn parse_args(args: env::Args) -> Result<(Args)> {
```

Also after seeing all of this arg parsing (it looks good), I wish I had recommended that we use `StructOpt`. But we don't have to do that now.